businessLogicLayer/Features/PricesNearMe/StationDataManager.swift

```diff
@@ -0,0 +1,123 @@
+                }
+            }
+        }
+    }
+    
+    // MARK: - Gathering data from StationDataAPI
+    
+    func gatherStationData(from endPointAddresses: [String], completion: @escaping (Result<[PetrolStation], StationDataManagerError>) -> Void) {
+        let stationDataAPI = self.stationDataAPI
+        let dispatchGroup = DispatchGroup()
+        var responses: [ResponseData] = []
+        
+        for (endPointAddress) in endPointAddresses {
+            guard let url = URL(string: endPointAddress) else {
+                print("Invalid URL for \(endPointAddress)")
+                continue
+            }
+            
+            dispatchGroup.enter()
+            
+            // let the API perform the network request. It shouldn't be aware of how to handle a list of urls. Instead, we simply make multiple calls using it. It's the role of that class to perform a network request and this class to loop through many urls. 😃
+            // Also it know means that we could use the stationDataAPI to make one single call if we wanted it to - it's even more flexible now.
+            stationDataAPI.requestStationData(from: url) { result in
+                defer {
+                    dispatchGroup.leave()
+                }
+                switch result {
+                case .success(let response):
+                    responses.append(response)
+                case .failure(let error):
+                    // TODO: handle this error, unless it's acceptable to ignore it.
+                    break
+                }
+            }
+        }
+        
+        dispatchGroup.notify(queue: .main) {
+            // TODO: We have not handled any of the errors passed to use from each network call.
+            guard responses.count > 0 else { // This is a hacky approach to error handling :(
+                completion(.failure(.refreshAlreadyInProgress))
+                return
+            }
+            // amend the responses to provide a list of all found stations.
+            let allStations: [PetrolStation] = responses.flatMap{ $0.stations }
+            
+            // Remove duplicate stations from array. This is efficient because we only iterate through the array once and it's done at the time of reciving the data too
+            var registeredSiteIDs: Set<String> = []
+            let stationsWithoutDuplicates: [PetrolStation] = allStations.compactMap { stationData in
+                guard registeredSiteIDs.contains(stationData.site_id) == false else {
+                    return nil // removes the duplicate from the array
+                }
+                registeredSiteIDs.insert(stationData.site_id)
+                return stationData
+            }
+            completion(.success(stationsWithoutDuplicates))
+        }
+    }
+    
+    // MARK: - Filtering via postcode
+    
+    private func filterStations(_ stations: [PetrolStation], forPostcode postcode: String) -> [PetrolStation] {
```

Comment:
This function now has no access to values outside of it. This is a common approach in Swift and also is considered a more "senior" way of doing things. It means that the function is more isolated and only manipulates data passed into it. Then it will return a value rather than setting one on the object its declared in.